### PR TITLE
Move X-Ray Env Variable propagation to span link instead of parent

### PR DESCRIPTION
**Pull request by tylerbenson**
https://github.com/open-telemetry/opentelemetry-specification/pull/3166

Per discussion in the FAAS SIG, we decided that the AWS X-Ray environment variable should be moved to a span link to avoid interfering with the configured propagators.

Also related to #5167.

### CODE REVIEW

File: instrumentation/aws-lambda/aws-lambda-core-1.0/library/src/main/java/io/opentelemetry/instrumentation/awslambdacore/v1_0/internal/AwsXRayEnvSpanLinksExtractor.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+      return;
+    }
+    Context parentCtx =
+        AwsXrayPropagator.getInstance()
+            .extract(
+                // see BaseTracer#extract() on why we're using root() here
+                Context.root(),
+                Collections.singletonMap(AWS_TRACE_HEADER_PROPAGATOR_KEY, parentTraceHeader),
+                MapGetter.INSTANCE);
+    SpanContext parent = Span.fromContext(parentCtx).getSpanContext();
```
Comment (tsloughter) [inline instrumentation/aws-lambda/aws-lambda-core-1.0/library/src/main/java/io/opentelemetry/instrumentation/awslambdacore/v1_0/internal/AwsXRayEnvSpanLinksExtractor.java:51]: Nit pick. It isn't a `parent`, so could be confusing for future readers to see `parentCtx` and `parent`.

Comment (tylerbenson) [reply instrumentation/aws-lambda/aws-lambda-core-1.0/library/src/main/java/io/opentelemetry/instrumentation/awslambdacore/v1_0/internal/AwsXRayEnvSpanLinksExtractor.java]: better now?

Comment (tsloughter) [reply instrumentation/aws-lambda/aws-lambda-core-1.0/library/src/main/java/io/opentelemetry/instrumentation/awslambdacore/v1_0/internal/AwsXRayEnvSpanLinksExtractor.java]: Yup :+1:

**Review by tsloughter — APPROVED** [review]

**Comment by laurit** [comment]
@tylerbenson Considering that all backends don't support span links would it make sense to make this configurable so vendors/users could revert to previous behaviour. WDYT?

**Comment by tylerbenson** [comment]
@laurit The only vendor that should care about the current behavior is Amazon/XRay. For everyone else, this behavior interferes with the regular trace propagation and results in broken traces. Moving it to a span link solves the broken propagation and maintains the information for x-ray customers.

File: instrumentation/aws-lambda/aws-lambda-core-1.0/library/src/main/java/io/opentelemetry/instrumentation/awslambdacore/v1_0/internal/AwsLambdaFunctionInstrumenter.java [hunk 2]
```diff
@@ -46,15 +47,25 @@ public void end(
...
+    INSTANCE;
+
+    @Override
+    public Iterable<String> keys(Map<String, String> map) {
+      return map.keySet();
+    }
+
+    @Override
+    public String get(Map<String, String> map, String s) {
+      return map.get(s.toLowerCase(Locale.ROOT));
```
Comment (laurit) [inline instrumentation/aws-lambda/aws-lambda-core-1.0/library/src/main/java/io/opentelemetry/instrumentation/awslambdacore/v1_0/internal/AwsLambdaFunctionInstrumenter.java:68]: why do you need to use `toLowerCase` here?

Comment (laurit) [reply instrumentation/aws-lambda/aws-lambda-core-1.0/library/src/main/java/io/opentelemetry/instrumentation/awslambdacore/v1_0/internal/AwsLambdaFunctionInstrumenter.java:68]: Because the headers map is converted to lower case https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/09577f97ffc0ca9f703386e12afbd018d12a2935/instrumentation/aws-lambda/aws-lambda-core-1.0/library/src/main/java/io/opentelemetry/instrumentation/awslambdacore/v1_0/internal/MapUtils.java#L18

File: instrumentation/aws-lambda/aws-lambda-core-1.0/library/src/main/java/io/opentelemetry/instrumentation/awslambdacore/v1_0/internal/AwsXrayEnvSpanLinksExtractor.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+  public static void extract(SpanLinksBuilder spanLinks) {
+    String parentTraceHeader = System.getenv(AWS_TRACE_HEADER_ENV_KEY);
+    if (parentTraceHeader == null || parentTraceHeader.isEmpty()) {
+      return;
+    }
+    Context xrayContext =
+        AwsXrayPropagator.getInstance()
+            .extract(
+                // see BaseTracer#extract() on why we're using root() here
```
Comment (mateuszrzeszutek) [inline instrumentation/aws-lambda/aws-lambda-core-1.0/library/src/main/java/io/opentelemetry/instrumentation/awslambdacore/v1_0/internal/AwsXrayEnvSpanLinksExtractor.java:52]: There no longer is a class named `BaseTracer` in this project; consider removing this comment.

File: instrumentation/aws-lambda/aws-lambda-core-1.0/library/src/main/java/io/opentelemetry/instrumentation/awslambdacore/v1_0/internal/AwsXrayEnvSpanLinksExtractor.java [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+import io.opentelemetry.instrumentation.awslambdacore.v1_0.AwsLambdaRequest;
+import java.util.Collections;
+import java.util.Locale;
+import java.util.Map;
+
+/**
+ * This class is internal and is hence not for public use. Its APIs are unstable and can change at
+ * any time.
+ */
+public final class AwsXrayEnvSpanLinksExtractor implements SpanLinksExtractor<AwsLambdaRequest> {
```
Comment (mateuszrzeszutek) [inline instrumentation/aws-lambda/aws-lambda-core-1.0/library/src/main/java/io/opentelemetry/instrumentation/awslambdacore/v1_0/internal/AwsXrayEnvSpanLinksExtractor.java:26]: Nit: can this class be package-private?

**Review by mateuszrzeszutek — APPROVED** [review]

**Comment by laurit** [comment]
@tylerbenson I understand what you are trying to do. I haven't used aws lambdas myself so perhaps I'm misunderstanding something. Let me rephrase my question. Our aws sdk instrumentation https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/instrumentation/aws-sdk/aws-sdk-2.2/library#trace-propagation currently uses only x-ray headers for propagation. If a request sent from aws sdk triggers a lambda that should only have the x-ray header. Previously parent context was set based on that x-ray header, now a span link is created instead. Is it possible that on a backend that does not support span links previously you got a complete trace and now after these changes you'll get a broken trace (because of missing support for span links)?

File: instrumentation/aws-lambda/aws-lambda-core-1.0/library/src/main/java/io/opentelemetry/instrumentation/awslambdacore/v1_0/internal/ParentContextExtractor.java [hunk 5]
```diff
@@ -1,81 +0,0 @@
...
-  static Context extract(Map<String, String> headers, AwsLambdaFunctionInstrumenter instrumenter) {
-    Context parentContext = null;
-    String parentTraceHeader = System.getenv(AWS_TRACE_HEADER_ENV_KEY);
-    if (parentTraceHeader != null) {
-      parentContext = fromXrayHeader(parentTraceHeader);
-    }
-    if (!isValidAndSampled(parentContext)) {
-      // try http
-      parentContext = fromHttpHeaders(headers, instrumenter);
-    }
```
Comment (tylerbenson) [inline instrumentation/aws-lambda/aws-lambda-core-1.0/library/src/main/java/io/opentelemetry/instrumentation/awslambdacore/v1_0/internal/ParentContextExtractor.java:36]: @laurit This is the particularly problematic section in the current releases.  Notice that it first uses the context from an environment variable if available, not the context from the incoming message.  This results in broken traces even if the x-ray context was sent via the aws-sdk as you linked to.  Rather than completely removing the environment variable provided context, this PR instead (and what is now required by the specification) adds that as a span link and restores regular context propagation from the incoming message.

Sorry, I should have explained better in my PR description.

**Review by laurit — APPROVED** [review]
